### ISSUE TITLE
Access typedefs from `Ontology` instance

### ISSUE THREAD

**Issue by vlmir** [althonos/pronto#19]
At least in my hands pronto imports no Typedef stanzas.
Is this a feature or a bug?

**Comment by althonos** [comment]
Actually they are extracted, and used to instantiate new relationships. They are not however accessible from the `Ontology` they were extracted from, which would be a nice feature.

**Comment by althonos** [comment]
As a test:
```python
import pronto
mondo = pronto.Ontology('http://purl.obolibrary.org/obo/mondo.obo')
print(mondo['MONDO:0000065'].relations)
```python
{Relationship('disease_arises_from_feature'): [<MONDO:0005015: diabetes mellitus (disease)>],
 Relationship('is_a'): [<MONDO:0005385 {source="DC:0000281"}: >,
                        <MONDO:0020573: genetic predisposition to disease>]}
```
(`disease_arises_from_feature` is typedefed in the `mondo.obo` file).

**Comment by vlmir** [comment]
Hi Martin,

thanks for your response. Unfortunately, it doesn't address the problem I
have, probably I was not clear enough about it. Let me re-state my
question.
As I understand it, correct me if I'm wrong, Pronto generates objects
corresponding only to the [Term] stanzas but not to the [Typedef] stanzas.
You see, my concern is Typedefs, not Relationships. Consequently, the
exports from Pronto are completely devoid of [Typedef] stanzas.
Moreover, the exported ontologies are effectively molded into taxonomies
built exclusively on subsumption - there are 'is_a:' statements but no
'relationship:' statements.
Thus, Pronto apparently fails to preserve the semantics present in the
input ontologies.
Have I missed something here? Is there any way to specify what should be
included in the exports?

Hope to hear from you soon,
Vladimir

P.S. I exported from Pronto like this:
 with open('test.obo', 'w') as fp:
   fp.write(onto.obo)

**Comment by althonos** [comment]
Hi Vladimir, 
I was talking about this, but I didn't think about the exported OBO, and indeed it will contain no typedefs, which is a problem.

What I think I'll do is adding a `typedefs` attribute to `Ontology` instances, containing only the typedefs defined in the ontology file. I will then be able to serialized only the referenced typedefs when exporting.

I don't think this will be too complicated to do, so I'll try to work on it this week-end.

**Comment by vlmir** [comment]
Hi Martin,

does this mean that Pronto, after all, parses Typedefs and retains them in
the ontology objects?
If so, then I'd like first to make clear that I have no vested interest in
the OBO export, I used it just as a sanity check.
I came across Pronto in a quest for a tool enabling Turtle/TTL
serialization of ontologies. That's my primary interest. Turtle is
gradually becoming a de facto standard language for Semantic Web
implementation, ousting luckily the original RDFXML standard.
My second interest is to have exports to JSON. Are Pronto objects
serializable via the Python 'json' module? If not, would it be possible for
you to implement a function transforming native objects into json
serializable ones like they do in the NetworkX library?

 Hope to hear from you soon,
Vladimir

**Comment by althonos** [comment]
It parses them but do not hold them in ontology objects, only as `Relationship` instances.

The Turtle parser will probably take more time since I'm not familiar with the format or the available parser libraries.

Exporting ontologies to JSON is already possible, though, using the `Ontology.json` property ! :wink: Although it's more of a dump of the internal data representation, and does not follow a specific JSON ontology format (such as *obojson*).

**Comment by vlmir** [comment]
Hi Martin,

what you say about json is already a good news. I personally don't care
much about what they contemplate about json at the OBO Foundry. The most
important thing for me is that I can get a json serialization fully
representative of the Ontology object.
Is the usage of the Ontology.json is anywhere documented?

Cheers,
Vladimir

**Comment by althonos** [comment]
It's listed on the Ontology reference (http://pronto.readthedocs.io/en/latest/pronto/pronto.Ontology.html#pronto.Ontology.json), but the documentation is not explanatory at all :blush:

Basically the JSON should contain all terms referenced by accession, as well as metadata.

**Comment by vlmir** [comment]
Thanks, Martin,
I'll have a look.
Vladimir

**Comment by puethe** [comment]
Hi there,
I just stumbled over the same issue. Have you managed to have a look at it yet?
I'm using Pronto to import data from ontology files into a database and basically can't establish relationships between terms before adding the corresponding typedefs to the database.
If you don't have the time, I can also offer to have a look at it myself and open a pull request. Let me know.
Cheers,
Christoph

**Comment by althonos** [comment]
@puethe: I added a `typedef` attribute to `Ontology` instances, and added a `obo` property to  the `Relationship` class so now exporting an ontology to `obo` format will export the `[Typedef]` stanzas, and you'll be able to access the typedefs declared in an ontology.

Could you test the master branch a bit to tell me if you need anything before I make a release ?

**Comment by puethe** [comment]
Thanks! That's very helpful and works for me.

Since the 'term' and 'typedef' stanzas are structurally very similar, at least in [obo](https://owlcollab.github.io/oboformat/doc/GO.format.obo-1_4.html), I personally probably would have created a base class from which both the 'Term' class and a similar 'Typedef' class inherit. That would allow catching the remaining properties of the typedefs, such as comments, definitions etc.

None of this, however, is crucial for me now, and might also just be my personal use case. So just go ahead and make a release if you like. Thanks again!

**Comment by althonos** [comment]
@puethe : I really wish I had more time to put into this library ! :wink:

**Comment by althonos** [comment]
Fixed in release `0.11.0`.
